src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,325 @@
+      temp_pad.CopyFromVec(signal_block_padded.Range(block_length, filter_length - 1));
+    } else {
+      signal->Range(po, signal_length - po).CopyFromVec(
+                        signal_block_padded.Range(0, signal_length - po));
+      if (filter_length - 1 < signal_length - po)
+        signal->Range(po, filter_length - 1).AddVec(1.0, temp_pad);
+      else
+        signal->Range(po, signal_length - po).AddVec(1.0, temp_pad.Range(0, signal_length - po));
+    }
+  }
+}
+
+void AddVectorsOfUnequalLength(const Vector<BaseFloat> &filter, Vector<BaseFloat> *signal) {
+  for (int po = 0; po < signal->Dim(); po += filter.Dim()) {
+    int block_length = filter.Dim();
+    if (signal->Dim() - po < block_length) block_length = signal->Dim() - po;
+    signal->Range(po, block_length).AddVec(1.0, filter.Range(0, block_length));
+  }
+}
+
+BaseFloat MaxAbsolute(const Vector<BaseFloat> &vector) {
+  return std::max(std::abs(vector.Max()), std::abs(vector.Min()));
+}
+
+BaseFloat ComputeEarlyReverbEnergy(const Vector<BaseFloat> &rir, const Vector<BaseFloat> &signal,
+                                   BaseFloat samp_freq) {
+      int peak_index = 0;
+      rir.Max(&peak_index);
+      KALDI_VLOG(1) << "peak index is " << peak_index;
+
+      const float sec_before_peak = 0.001;
+      const float sec_after_peak = 0.05;
+      int early_rir_start_index = peak_index - sec_before_peak * samp_freq;
+      int early_rir_end_index = peak_index + sec_after_peak * samp_freq;
+      if (early_rir_start_index < 0) early_rir_start_index = 0;
+      if (early_rir_end_index > rir.Dim()) early_rir_end_index = rir.Dim();
+
+      int duration = early_rir_end_index - early_rir_start_index;
+      Vector<BaseFloat> early_rir(rir.Range(early_rir_start_index, duration));
+      Vector<BaseFloat> early_reverb(signal);
+      BlockConvolveSignals(early_rir, &early_reverb);
+
+      // compute the energy
+      return VecVec(early_reverb, early_reverb) / early_reverb.Dim();
+}
+
+void UnitTestBlockConvolution(const Vector<BaseFloat> &rir, Vector<BaseFloat> *signal) {
+    Vector<BaseFloat> signal_test(*signal);
+    ConvolveSignals(rir, &signal_test);
+    BlockConvolveSignals(rir, signal);
+    AssertEqual(*signal, signal_test, 0.001 * signal->Dim());
+}
+
+void DoReverberation(const Vector<BaseFloat> &rir, Vector<BaseFloat> *noise,
+                        Vector<BaseFloat> *signal, BaseFloat samp_freq,
+                        BaseFloat snr_db, bool test_mode) {
+  if (noise->Dim()) {
+    float input_power = ComputeEarlyReverbEnergy(rir, *signal, samp_freq);
+    float noise_power = VecVec(*noise, *noise) / noise->Dim();
+    float scale_factor = sqrt(pow(10, -snr_db / 10) * input_power / noise_power);
+    noise->Scale(scale_factor);
+    KALDI_VLOG(1) << "scale_factor of channel: " << scale_factor << " for SNR db " << snr_db;
+  }
+
+  if (test_mode) {
+    UnitTestBlockConvolution(rir, signal);
+  } else {
+    BlockConvolveSignals(rir, signal);
+  }
```

Comment:
Second @vijayaditya comment, please separate into a library file and unit-test it separately from the utility. As is it will not be run by automated `make test` from src/. @dpovey, what would be the best library for signal-processing primitives? `src/feat`?